plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -404,12 +403,156 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize, b
+            final int nbParts = multiParts.v1().intValue();
+            final long lastPartSize = multiParts.v2();
+            assert blobSize == (((nbParts - 1) * partSize) + lastPartSize) : "blobSize does not match multipart sizes";
+
+            final List<String> blockIds = new ArrayList<>(nbParts);
+            for (int i = 0; i < nbParts; i++) {
+                final long length = i < nbParts - 1 ? partSize : lastPartSize;
+                final Flux<ByteBuffer> byteBufferFlux =
+                    convertStreamToByteBuffer(inputStream, length, DEFAULT_UPLOAD_BUFFERS_SIZE);
+
+                final String blockId = UUIDs.base64UUID();
+                blockBlobAsyncClient.stageBlock(blockId, byteBufferFlux, length).block();
+                blockIds.add(blockId);
+            }
+
+            blockBlobAsyncClient.commitBlockList(blockIds, failIfAlreadyExists == false).block();
+        });
+    }
+
+    /**
+     * Converts the provided input stream into a Flux of ByteBuffer. To avoid having large amounts of outstanding
+     * memory this Flux reads the InputStream into ByteBuffers of {@code chunkSize} size.
+     * @param inputStream the InputStream to convert
+     * @param length the InputStream length
+     * @param chunkSize the chunk size in bytes
+     * @return a Flux of ByteBuffers
+     */
+    private Flux<ByteBuffer> convertStreamToByteBuffer(InputStream inputStream, long length, int chunkSize) {
+        assert inputStream.markSupported() : "An InputStream with mark support was expected";
+        // We need to mark the InputStream as it's possible that we need to retry for the same chunk
+        inputStream.mark(Integer.MAX_VALUE);
+        return Flux.defer(() -> {
+            final AtomicLong currentTotalLength = new AtomicLong(0);
+            try {
+                inputStream.reset();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            // This flux is subscribed by a downstream operator that finally queues the
+            // buffers into netty output queue. Sadly we are not able to get a signal once
+            // the buffer has been flushed, so we have to allocate those and let the GC to
+            // reclaim them (see MonoSendMany). Additionally, that very same operator requests
+            // 128 elements (that's hardcoded) once it's subscribed (later on, it requests
+            // by 64 elements), that's why we provide 64kb buffers.
+            return Flux.range(0, (int) Math.ceil((double) length / (double) chunkSize))
+                .map(i -> i * chunkSize)
+                .concatMap(pos -> Mono.fromCallable(() -> {
+                    long count = pos + chunkSize > length ? length - pos : chunkSize;
+                    int numOfBytesRead = 0;
+                    int offset = 0;
+                    int len = (int) count;
+                    final byte[] buffer = new byte[len];
```

Comment:
This still scares me a little. If we only do 64k at a time here, can't we use the Netty memory allocator (or manage  our own set of `byte[]` and recycle them on `doOnComplete` or do we still have no guarantees about flushing at that point?

Reply:
Sadly we don't have guarantees about flushing at that point

Reply:
I explored a different approach where I was passing an allocator there, and recycling at the end of the request, but in that case you end up holding that memory for the entire duration of the request.